[#122189093] Fix ctl setup

mtekel opened the pull request:
### What

As it turns out, ctl_setup script is adding all the libraries of all packages deployed on a VM by BOSH into `LD_LIBRARY_PATH`. In case of concourse, this includes various resources, which are based on busybox. This then breaks commands like `mkdir` (segfault) - which breaks the start script in the end and monit is not able to run collectd, failing the VM deployment.

Move directory creation before `LD_LIBRARY_PATH` is exported. This way it works regardless of what packages are present on the system.
### Testing
- to verify that this doesn't break existing installation, switch your pipeline to deploy paas-collectd-bosh release from branch of this PR, deploy & check you still get metrics:

```
diff --git a/concourse/pipelines/create-bosh-cloudfoundry.yml b/concourse/pipelines/create-bosh-cloudfoundry.yml
index 723840e..7ff1a26 100644
--- a/concourse/pipelines/create-bosh-cloudfoundry.yml
+++ b/concourse/pipelines/create-bosh-cloudfoundry.yml
@@ -110,7 +110,10 @@ resources:
     type: git
     source:
       uri: https://github.com/alphagov/paas-collectd-boshrelease.git
-      tag_filter: {{cf_collectd_version}}
+      #tag_filter: {{cf_collectd_version}}
+      #FIXME: once this release has been tested remove the line below
+      # and uncomment the one above
+      branch: fix-ctl-setup

   - name: os-conf-boshrelease
     type: git
diff --git a/manifests/cf-manifest/manifest/runtime/runtime.yml b/manifests/cf-manifest/manifest/runtime/runtime.yml
index 3d5d8be..9551913 100644
--- a/manifests/cf-manifest/manifest/runtime/runtime.yml
+++ b/manifests/cf-manifest/manifest/runtime/runtime.yml
@@ -2,7 +2,10 @@ releases:
   - name: os-conf
     version: commit-a2bc2ab32248c8edf7c4790b33902893b1f4db66
   - name: collectd
-    version: 0.3
+    #version: 0.3
+    #FIXME: once this release has been tested remove the line below
+    # and uncomment the one above
+    version: test-version-1

 addons:
   - name: os-configuration
```
- to verify that this fixes startup issue, deploy this on concourse (see https://github.com/alphagov/paas-cf/pull/361), then downgrade to 0.3 version and see it fail:

```
  - name: collectd
    sha1: cd58286ff328d3dcd51507a434f010225e5179ca
    url: https://github.com/alphagov/paas-collectd-boshrelease/releases/download/0.3/collectd-0.3.tgz
```
### After merge

Tag master as 0.4. Create a 0.4 release here on github and add the tarball to it.
### Who

not @mtekel or @HenryTK

keymon commented:
Instead of this, I propose filter out the packages that contain the `ld-*.so` library, as they are likely a rootfs (like the resources):

```
for package_bin_dir in  /var/vcap/packages/*/lib
do
    # do not include a package with ld-*.so as it is likely a rootfs
    if [ -z "$(find ${package_bin_dir} -name 'laaad-*.so')" ]; then
        export LD_LIBRARY_PATH=${package_bin_dir}:$LD_LIBRARY_PATH
    fi
done 

``` 

I tested it and it seems to work fine

mtekel commented:
We thought about this, but the idea is to not impact mkdir with _any_ special/different libraries. Thus move it in the beginning of the script. Then we'd follow up with the author to find out the real reason behind this greedy include of all libraries into `LD_LIBRARY_PATH` . We have also tested excluding `_resource` in the list of dirs libraries are searched in - and that worked as well. But we did not want to go the way of various custom modifications of which libraries should be loaded for mkdir...

keymon commented:
I think it is pretty safe to think that any package containing `ld*.so` will fail and not work, and we should do a PR upstream proposing this change.

I wouldn't use `*-resource` or even that `busybox` exclusions, because they are too specific to a given set of packages in a given installation.

mtekel commented:
Nevertheless, in both cases, doing directory creation first avoids the problem. We can raise the rest in the upstream PR and wait for the reaction of the author...

keymon commented:
It avoids the mkdir from crashing because it does not run so the script works. But obviously the script shouldn't set `LD_LIBRARY_PATH` with a bogus value that might make any process crash in any minute.

This can bite us later, so I would rather set `LD_LIBRARY_PATH` with a safer value than reorder the script and fix it now that we have the context.

mtekel commented:
I would re-order the script & fix it now. Raise a PR to author for finding out why they load the libraries and if our fix is acceptable. If merged, we can backport to here. In any case, I don't see a reason why you would want to run mkdir with a specific LD_LIBRARY_PATH.

keymon commented:
Yep +1 for doing both.

As a side comment, that script is quite unsafe to run in general, it just assumes that you can merge all the paths for python, java, etc... just because they might be required for some plugin in collectd. A better approach should be prefix create custom packages with a name like `collect-dep-*`... but that is a different story and it is out of scope.

keymon commented:
Let me know when you change this.

keymon commented:
I think change the order of  the script to have `mkdir` before is not a complete solution. It seems to work but we cannot be sure it will fail later.

This failure highlighted that  `LD_LIBRARY_PATH` is being set with paths that potentially can make break any program, as in concourse is loading the `/lib` of resources which includes `ld-*.so` and `libc*.so` which are, obviously, not intended to be loaded with collectd. 

There is a proposed easy solution: If we filter out any package which contains a `libc` or `ld.so`, we will avoid include libraries from packages which include a rootfs (resources, linux-rootfs for cells...)

IMO, we should not t merge this PR until the we fix the root cause of the problem (bogus LD_LIBRARY_PATH). There should be a solution implemented, either different PR or in this same PR. I think it should be one same PR because moving `mkdir` simply hides the issue.

I don't think this should be merged in the pass-cf PR https://github.com/alphagov/paas-cf/pull/361 because it is a ticking bomb. The LD_LIBRARY_PATH is still bogus, loads all,  in any moment it will fail: any process might die randomly or if we add, change or update any resource.

alext commented:
Another point to raise: Currently the `0.4` tag exists and points to the commit in this PR (not a commit on master). This doesn't seem right. This tag should have been created after this was merged, and be pointed at the merge commit.

keymon commented:
I asked @simonjohansson in person, no special reason to implement this the way they did, they only needed to load LD dependencies for some collectd plugins (psql, mysql, etc).

mtekel commented:
Thanks @keymon , that's useful kind of feedback. Based on this I'll move LD_LIBRARY_PATH setting to the end of the script so that it doesn't impact anything in the script and add the filter you suggested to ignore system level libraries.

mtekel committed:
CTL_SETUP: Set PATH and LD_LIBRARY_PATH as last

Only configure PATH and LD_LIBRARY_PATH as last actions in the script.
This way it doesn't affect commands of the script (except the last echo).

mtekel committed:
CTL_SETUP: filter likely rootfs libs

Don't include libraries from packages that look like rootfs in to the
LD_LIBRARY_PATH. This can happen when you are deploying collectd to e.g.
concourse server which contains busybox based resources as packages.

Busybox libraries are not compatible with stemcell based system commands
/packages and break things if included in the LD_LIBRARY_PATH.

mtekel commented:
I re-did the commits. Created a new release: https://s3-eu-west-1.amazonaws.com/mt-state/collectd-0.4.tgz

I have removed the 0.4 release draft and deleted 0.4 tag. We should only tag & create 0.4 release when this is merged (add the above linked tarball to the release or you can build it by `bosh create release --name collectd --with-tarball --force --version 0.4`)

mtekel committed:
CTL_SETUP: append to path

When you pre-pend to path and have local packages that include basic OS
commands, like sleep, those will get used in the scripts (like when
stopping collectd with monit) instead of system sleep command. It can then
happen that these package specific system level commands don't support
all the same parameters (like 0.1 time for sleep). This then breaks the
scripts.

Append to PATH instead. This will ensure that stemcell system level scripts
will get used first, but any specific binaries from packages will still be
included in the path.

mtekel commented:
Closing in favor of #2 - that PR is against gds_master (as it should be).
